### PR TITLE
PoX-4 Signing Keys Stacking-State

### PR DESCRIPTION
**Pull request by setzeus**
# PoX-4 Signing Key Stacking-State Update
## Summary of Changes

This PR is meant to introduce, not complete, the core logic around signing keys in PoX-4 & address #4057 . The focus is specifically on updating the 'stacking-state' map & all the functions that read & write from it. This includes updates to the following public functions:

1. (stack-stx)
2. (delegate-stack-stx)
3. (stack-extend)
4. (delegate-stack-extend)

_what this does not include_
Getters & setters to the new map 'reward-cycle-signing-keys,' that'll be in another PR. This PR also does not include any logic that'll be used in relevant issues #4058, #4059, & #3970.

### How were these changes tested?

This was only tested locally for syntax errors / clarinet check.

### What future testing should occur?
As soon as this is merged I'll open an issue with the user stories that require testing for @moodmosaic to work into pox_4_tests.rs.

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by setzeus** [comment]
Additional context re: using 'delegate-stack-stx' vs. 'stack-aggregation-commit-index' to submit a pool's signing-key for delegates that registered without a signing-key here:

https://github.com/stacks-network/stacks-core/discussions/4082

**Comment by setzeus** [comment]
Note: per Clarity WG morning discussion it's been noted that our preferred workflow here is to actively work on a version of PoX-4 that exists in /contrib per #4094 & then periodically copy that over to the more hardened PoX-4 that'll live in /boot. 

Once #4094 is merged in I'll make these adjust *only* to the PoX-4 version in /contrib for now.

(unless there's reasonable push back on this workflow)

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:601]: Typo -- extra parenthesis.

Please make sure that this contract instantiates.  The reason your tests are failing is because there are syntax errors in `pox-4`, which prevents it from instantiating.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:658]: Please delete this comment

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:1039]: can you fix this typo? `etry`

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:1043]: This error case shouldn't be reachable because `stack-stx` requires a signing key.  Maybe `unwrap-panic` would be better here?

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:1221]: Same -- `unwrap-panic` might be a better choice here.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (kantai) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:123]: It seems like this could just be `(buff 33)` rather than an `(optional (buff 33))`: the delegate sets this value in `delegate-stack-stx`, which is also the first time the stacker-state is instantiated for the delegator.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (kantai) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:150]: I'm curious about this map: what is it used for?

**Review by kantai — COMMENTED** [review]
Echoing Jude's comments, and just had a few more

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (friedger) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:123]: Does it make sense to allow delegators enforce a signing key? If yes, should it be stored in a separate map?

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (kantai) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:150]: Never mind -- I see it's discussed here #4059

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (friedger) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:123]: Sorry, I meant the user/pool member. My question does not make sense here.

Comment (friedger) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:123]: I agree, that stack-aggregation-commit can't change the signing key.

**Review by friedger — CHANGES_REQUESTED** [review]
I try to understand whether the same logic for the pox-address should be applied for the signing key for pools: If the pool operator decides to use a different signer key, the pool members rewards are not really at risk. However, would pool members like to decide whether the pool operator can delegate to one or the other signer? I think yes, therefore `delegate-stx` should also contain a signing key in the delegation-state

**Comment by friedger** [comment]
@setzeus There at least one rust functions that create stack-stx transactions with 4 instead of 5 arguments that makes the unit tests fail.
I found one in blocks.rs: https://github.com/stacks-network/stacks-blockchain/blob/cef8f87c59d326b5d30d9f2e6baf258d820da6b1/stackslib/src/chainstate/stacks/db/blocks.rs#L4222

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:147]: Could you please add a getter?

Comment (moodmosaic) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:147]: @setzeus what `nextest` command you're running? Can you paste it here so I can look into it?

**Review by kantai — APPROVED** [review]

**Review by friedger — APPROVED** [review]

**Comment by muneeb-ali** [comment]
Great to see the changes approved here! 🚀

File: stackslib/src/chainstate/nakamoto/coordinator/tests.rs
Comment (jcnelson) [inline stackslib/src/chainstate/nakamoto/coordinator/tests.rs:286]: Please remove dead code

File: stackslib/src/chainstate/nakamoto/coordinator/tests.rs
Comment (jcnelson) [inline stackslib/src/chainstate/nakamoto/coordinator/tests.rs:288]: Why is this a new key?  Can you continue to please use the same key so we get consistent output?

File: stackslib/src/chainstate/stacks/boot/contract_tests.rs
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/contract_tests.rs:572]: What is this test supposed to check?  If all it's checking is that pox-4 is syntax-free, then these steps aren't needed -- you can go straight to instantiating pox-4.  If it's instead checking that pox-4 instantiates at the right time, then this test needs to verify that instead.

**Review by jcnelson — CHANGES_REQUESTED** [review]
Thank you for updating the stacking interfaces with these new methods!  Before merging, I must insist that you add unit tests for them.  It's not enough to see that PoX-4 instantiates; you must also verify that the public keys get added to the stacking state.  Thanks!

File: stackslib/src/chainstate/stacks/boot/contract_tests.rs
Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/contract_tests.rs:572]: Yes; however, this is not being tested here

Comment (moodmosaic) [reply stackslib/src/chainstate/stacks/boot/contract_tests.rs:572]: Then would a pox lockup transition from pox-3 to pox-4 be enough? (As a test.)

Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/contract_tests.rs:572]: Yes; you'd need to verify that the `instantiate_epoch_2_5()` and `initialize_epoch_3_0()` functions in `stackslib/src/clarity_vm/clarity.rs` are getting called.

Comment (moodmosaic) [reply stackslib/src/chainstate/stacks/boot/contract_tests.rs:572]: OK, and then this test shouldn’t be in this file but instead it should be in `pox_4_tests.rs` yes?

Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/contract_tests.rs:572]: Either way is fine with me. If @jbencin already tests this in his PR, then it's fine to omit it here.  However, it's really important that you test the new stacking API.

**Comment by jcnelson** [comment]
An update from the testing WG.  This PR is blocked on getting the `pox_4.rs` tests in place, which is being handled in https://github.com/stacks-network/stacks-core/pull/4106.  Once this is in place, it should be merged into this PR, so then it will become possible to test that the signing keys were indeed set.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:1228]: maybe simplify to
`(next-signing-key (defaults-to (get signing-key stacker-state) updated-signing-key))`

**Review by friedger — COMMENTED** [review]
nit on pox-4.clar

File: stackslib/src/chainstate/stacks/boot/mod.rs
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/mod.rs:1663]: This `key` is the key used to sign the transaction, not the DKG signing key.  There needs to be a separate argument to `make_pox_4_lockup` for that signing key.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:122]: I don't think this is ever `none` if we're setting the signing key on `delegate-stack-stx` (which IIRC is the plan now)

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:609]: Can you add a comment about `signing-key` to the method doc?

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:848]: Same here -- can you add a comment to the method doc about the signing key?

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:1050]: Agreed

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:1228]: Agreed

File: testnet/stacks-node/src/mockamoto.rs
Comment (jcnelson) [inline testnet/stacks-node/src/mockamoto.rs:827]: I think there instead needs to be a separate `signing_key` field in `MockamotoNode`.  Same for the Nakamoto node.  Perhaps it can be part of the `Keychain`?

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (setzeus) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:610]: Updated a sentence & added the following to comment doc above function:

`;; * You need to provide a signing-key used to allocate signature-weight to a valid signer`

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (setzeus) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:850]: Updated a sentence & added the following to comment doc above function:

`;; * You need to provide a signing-key used to allocate signature-weight to a valid signer`

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (moodmosaic) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:30]: Where is this constant used?

**Review by moodmosaic — COMMENTED** [review]
It would be nice with some tests to ensure everything works as intended:

1. **`signing-key` in stacking-state:**
   - Verify that the `signing-key` is correctly set.

2. **Modified Functions (`stack-stx`, `delegate-stack-stx`, etc.):**
   - Ensure all functions modified to include the `signing-key` handle it correctly.

3. **`stack-extend`, `delegate-stack-extend`:**
   - Test handling of `updated-signing-key`, ensuring proper updates in the stacking state.

4. **`get-signer-info-for-cycle`:**
   - Verify retrieval of correct signer info for various inputs.

@jcnelson, @setzeus, could you share your thoughts on these test cases or suggest additional ones that might be relevant?

**Comment by jcnelson** [comment]
@moodmosaic Yes, please test all of those in `pox_4_tests.rs`.

File: contrib/core-contract-tests/tests/bns/name_register_test.ts
Comment (jcnelson) [inline contrib/core-contract-tests/tests/bns/name_register_test.ts:101]: This looks like there was an off-by-one error somewhere, with all these changes. Can you elaborate more on why all of these block heights are now +1?

File: stackslib/src/chainstate/stacks/boot/mod.rs
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/mod.rs:1797]: Don't pass `None`; instead, make this an argument in Rust which gets passed to `stack-extend`.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:41]: Please don't commit commented-out code

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:37]: If you're now using `is-in-mainnet` instead of a separate `pox-4-mainnet.clar`, then where are these constants defined now?

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:540]: @setzeus This needs to be in #4164

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:584]: Please revert this -- it has not been decided yet that stacking will be disabled in the prepare phase

Comment (setzeus) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:584]: Yes. Reverting all the work here to un-do rebasing from #4094.

File: contrib/core-contract-tests/tests/bns/name_register_test.ts
Comment (friedger) [reply contrib/core-contract-tests/tests/bns/name_register_test.ts]: The pox-4 is in a different epoch than bns, therefore booting needs one more block

**Comment by setzeus** [comment]
Work picked up by @MarvinJanssen in #4209 to add required tests in addition to base tests - changing to draft until #4209 is merged in then will close.

**Comment by setzeus** [comment]
Closing since #4209 is now merged in.

**Pull request closed without merging**
